### PR TITLE
Refactor Challenge edit page 

### PR DESCRIPTION
**Pull request by piotrekpasciak**

### CODE REVIEW

File: app/controllers/challenges_controller.rb [hunk 1]
```diff
@@ -120,151 +106,8 @@ def set_challenge
...
-          ],
-          invitations_attributes:                 [
-            :id,
-            :email,
-            :_destroy
-          ])
-  end
-
   def set_organizer
-    @organizer = Organizer.friendly.find(params[:organizer_id])
```
Comment (nemaniarjun) [inline app/controllers/challenges_controller.rb]: Not sure about this change. What happens when an admin opens the challenge edit page?
Admin participant wont have organizer association. 
There is no point in doing this like that. Also in the future we will want to allow participants to be part of multiple organizations.

File: app/policies/challenge_policy.rb [hunk 2]
```diff
@@ -24,11 +24,11 @@ def update?
   end
 
   def new?
-    edit?
+    participant&.admin?
```
Comment (nemaniarjun) [inline app/policies/challenge_policy.rb]: what is the reason behind this change? the `edit` function also takes care of the participants who are not admin but are organizers.

File: app/views/challenges/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  <div class="container-fluid">
+    <div class="row">
+      <div class="col-12">
+        <%= form_with model: @challenge, class: 'form-rg challenges-form', multipart: true, local: true do |f| %>
+          <p id="challenges-form-errors" class="challenges-form__errors-paragraph"><%= @challenge.errors.full_messages.to_sentence %></p>
+
+          <% hidden_field_tag :step, id: 'challenges-form-step', value: :overview %>
+
+          <div class='tab-content'>
+            <div id='tab-details' class="tab-pane fade show active">
```
Comment (nemaniarjun) [inline app/views/challenges/_form.html.erb]: Will it be better to have an array of partials here which we can loop over? Would make this code more readable and allow for changes in the future.

File: app/controllers/challenges_controller.rb (hunk 1 above)
Comment (piotrekpasciak) [reply app/controllers/challenges_controller.rb]: That's a great catch 👍 , should be `challenge.organizer.` I've decided to drop organizer dependency for challenge.

File: app/policies/challenge_policy.rb (hunk 2 above)
Comment (piotrekpasciak) [reply app/policies/challenge_policy.rb]: I wanted to make sure that only admins can create new challenges, as this feature is not meant to be released.
 Previously there was no possibility to create new challenge through a form. Correct me if I'm wrong.

File: app/controllers/challenges_controller.rb (hunk 1 above)
Comment (nemaniarjun) [reply app/controllers/challenges_controller.rb]: Yes I agree with dropping the dependency for organizer for new challenges.

File: app/views/challenges/_form.html.erb (hunk 3 above)
Comment (piotrekpasciak) [reply app/views/challenges/_form.html.erb]: I disagree. I really feel this is so mutch easier to read thann version with loop and some kind of constants.

File: app/policies/challenge_policy.rb (hunk 2 above)
Comment (nemaniarjun) [reply app/policies/challenge_policy.rb]: I think there was. The organizers page (ex: https://www.aicrowd.com/organizers/imageclef) has a new challenge button which allows them to create a new challenge. The changing of the status from :draft to any other was something that only admins are allowed to do, but an organizer should be able to draft a challenge all by themselves.

File: app/views/challenges/_form.html.erb (hunk 3 above)
Comment (nemaniarjun) [reply app/views/challenges/_form.html.erb]: Hmm, I used to think the same way, but after what happened with the challenge _subnav partial, I think we have to move to loops at some point to make things easier. 
For now of course we can leave this like it is.

File: app/views/challenges/_navigation.html.erb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<div class="sub-nav-bar sticky-top">
+  <div class="container-fluid">
+    <ul class='nav'>
+      <li class="nav-item"><a class="nav-link active" data-toggle='tab' href="#tab-details" onclick="return false;">Details</a></li>
+      <% if challenge.persisted? %>
+        <li class="nav-item"><a class="nav-link" data-toggle='tab' href="#tab-overview" onclick="return false;">Overview</a></li>
+        <li class="nav-item"><a class="nav-link" data-toggle='tab' href="#tab-type" onclick="return false;">Rounds</a></li>
```
Comment (nemaniarjun) [inline app/views/challenges/_navigation.html.erb]: seems like a typo. Should be `tab-rounds` instead of `tab-type`

File: app/policies/challenge_policy.rb (hunk 2 above)
Comment (piotrekpasciak) [reply app/policies/challenge_policy.rb]: @nemaniarjun I've added commit that allows admins and organizers to create new challenge.

File: app/views/challenges/_form.html.erb (hunk 3 above)
Comment (piotrekpasciak) [reply app/views/challenges/_form.html.erb]: I prefer to leave like that for now.

File: app/views/challenges/_navigation.html.erb (hunk 4 above)
Comment (piotrekpasciak) [reply app/views/challenges/_navigation.html.erb]: Thanks, that's a good catch.

File: app/models/challenge.rb [hunk 5]
```diff
@@ -73,7 +72,6 @@ class Challenge < ApplicationRecord
 
   validates :status, presence: true
   validates :challenge, presence: true
-  validates :organizer_id, presence: true
```
Comment (nemaniarjun) [inline app/models/challenge.rb]: Hmm this seems weird. 
Every challenge has to have an organizer linked to it right ?

File: app/controllers/challenges_controller.rb [hunk 6]
```diff
@@ -108,6 +110,11 @@ def set_organizer
     @organizer = @challenge&.organizer
   end
 
+  def set_organizers_for_select
+    # We'll need refactor to select2 with API endpoint when we'll have a lot of organizers.
+    @organizers_for_select = Organizer.limit(1000).pluck(:organizer, :id)
```
Comment (nemaniarjun) [inline app/controllers/challenges_controller.rb]: Organizer.all.pluck would be better right? we dont want to suddenly run into a bug due to this hardcoded number

File: app/models/challenge.rb (hunk 5 above)
Comment (piotrekpasciak) [reply app/models/challenge.rb]: I've deleted it, because this code is redundant, because of following line:
https://github.com/AIcrowd/AIcrowd/blob/master/app/models/challenge.rb#L8

`belongs_to :organizer` implements presence validation since Rails 5. To disable this validation on `belongs_to` you have to pass `optional: true` parameter.

Having 2 validations that do same thing has consequence of that we have 2 validation error message about same issue, what doesn't seem right.

File: app/controllers/challenges_controller.rb (hunk 6 above)
Comment (piotrekpasciak) [reply app/controllers/challenges_controller.rb]: I've added `1000` here because  usually at point  of 1000 records it makes sense to refactor to select2 with ajax to resolve . performance issues.

I'll implement your suggestion as it's actually seems more reasonable.

File: app/controllers/challenges_controller.rb [hunk 7]
```diff
@@ -55,37 +41,46 @@ def show
...
   end
 
-  def edit; end
-
   def create
-    @challenge                = @organizer.challenges.new(challenge_params)
-    @challenge.clef_challenge = true if @organizer.clef_organizer
+    @challenge                = Challenge.new(challenge_params)
+    @challenge.organizer_id   = current_participant.organizer_id if @challenge.organizer_id.blank?
+    @challenge.clef_challenge = true if @organizer&.clef_organizer
```
Comment (nemaniarjun) [inline app/controllers/challenges_controller.rb]: hmm.. if we are setting the organizer_id on line 50, then we should remove set_organizer before create and use this id for the organizer right?

Comment (piotrekpasciak) [reply app/controllers/challenges_controller.rb]: Currently this line is necessary to allow organizer to create a challenge. I'll check if I can drop it by modyfing the following input:
https://github.com/AIcrowd/AIcrowd/pull/1091/files#diff-08ced1b1da67c79e061e958b17f3ea61R147

Comment (nemaniarjun) [reply app/controllers/challenges_controller.rb]: not sure about this. 
The workflow I imagine
1. no set_organizer is called before create, its only called before new, edit and update `@organizer = @challenge&.organizer`
2. In the form using the input that you mentioned we will set the :organizer_id 
```
<% if @challenge.new_record? && current_participant&.admin? %>
      <%= f.select :organizer_id, @organizers_for_select, { include_blank: true } %>
    <% else %>
      <input class="form-control" type='text' value='<%= @challenge.organizer&.organizer || current_participant.organizer&.organizer %>' readonly>
    <% end %>
```
3. In create we use this id that was set to :organizer_id and use that for line 51
`@challenge.clef_challenge = true if @challenge.organizer&.clef_organizer`
or 
`@challenge.clef_challenge = true if Organizer.find(@challenge.organizer_id)&.clef_organizer`

Comment (nemaniarjun) [reply app/controllers/challenges_controller.rb]: Ah I just noticed that the readonly is the text part. we should also have an organizer_id field there which is readonly and set using `@challenge.organizer&.id || current_participant.organizer&.id`

Comment (nemaniarjun) [reply app/controllers/challenges_controller.rb]: On third thought, why are we setting challenge.clef_challenge to true? maybe we should make it a model attribute which just calls self.organizer.clef_organizer

Comment (piotrekpasciak) [reply app/controllers/challenges_controller.rb]: https://github.com/AIcrowd/AIcrowd/pull/1091#discussion_r371818837

Comment (piotrekpasciak) [reply app/controllers/challenges_controller.rb]: https://github.com/AIcrowd/AIcrowd/pull/1091#discussion_r371819475
I don't know why.  This is old code and I'm no familiar with business logic. Decided to not touch it.

Comment (nemaniarjun) [reply app/controllers/challenges_controller.rb]: Another idea is to have a select field for when an participant creates the challenge with a select field for all the organizations that the participant is a part of. 
We want to change the one-to-one mapping of participant as organizer to multiple allowing a participant to be a part of multiple organization

Comment (piotrekpasciak) [reply app/controllers/challenges_controller.rb]: https://github.com/AIcrowd/AIcrowd/pull/1091#discussion_r371817874 I've implement this remark.

Comment (piotrekpasciak) [reply app/controllers/challenges_controller.rb]: https://github.com/AIcrowd/AIcrowd/pull/1091#discussion_r371821357 - I think this is additional work. Don't want to add additional work to this PR. If this is needed I prefer to due it on separate branch.

File: app/controllers/challenges_controller.rb [hunk 8]
```diff
@@ -55,37 +41,45 @@ def show
   end
 
   def new
-    @challenge = @organizer.challenges.new
+    @challenge = Challenge.new(organizer: @organizer)
```
Comment (nemaniarjun) [inline app/controllers/challenges_controller.rb]: is this @organizer set now? :thinking: :thought_balloon:

Comment (piotrekpasciak) [reply app/controllers/challenges_controller.rb]: @nemaniarjun  This is not needed anymore. I'll remove it.

Comment (piotrekpasciak) [reply app/controllers/challenges_controller.rb]: @organizer is no longer set after previous changes

File: app/policies/challenge_policy.rb [hunk 9]
```diff
@@ -24,7 +24,7 @@ def update?
   end
 
   def new?
-    edit?
+    participant && (participant.admin? || @participant.organizer_id.present?)
```
Comment (nemaniarjun) [inline app/policies/challenge_policy.rb]: is it `participant` or `@participant`?

Comment (piotrekpasciak) [reply app/policies/challenge_policy.rb]: Both point to the same thing, I'll update it.

File: app/assets/stylesheets/challenges/form.scss [hunk 10]
```diff
@@ -0,0 +1,11 @@
+.challenges-form {
```
Comment (nemaniarjun) [inline app/assets/stylesheets/challenges/form.scss]: Okay, one last thing. Can you move these css changes to a separate commit? that will help later when we work with jason to update the styles. its not required but having a different commit will probably work in our favour for finding and changing this later.

Comment (piotrekpasciak) [reply app/assets/stylesheets/challenges/form.scss]: I've moved css styles to separate commit.

File: app/views/challenges/form/_form.html.erb [hunk 11]
```diff
@@ -1,78 +0,0 @@
...
-          <div id='tab-overview' class="tab-pane fade">
-            <%= render partial: 'challenges/form/overview', locals: { challenge: challenge } %>
-          </div>
-
-          <div id='tab-private' class="tab-pane fade">
-            <%= render partial: 'challenges/form/private_challenge', locals: { challenge: challenge } %>
-          </div>
-
-          <div id='tab-dataset' class="tab-pane fade">
-            <%= render partial: 'challenges/form/dataset', locals: { challenge: challenge } %>
```
Comment (nemaniarjun) [inline app/views/challenges/form/_form.html.erb]: @piotrekpasciak

Comment (piotrekpasciak) [reply app/views/challenges/form/_form.html.erb]: @nemaniarjun After refactor we no longer render it.
